storm-client/src/jvm/org/apache/storm/daemon/worker/Worker.java

```diff
@@ -299,6 +313,33 @@ public void doExecutorHeartbeats() {
         }
     }
 
+    public Map<String, Long> getCurrentBlobVersions() throws IOException {
+        Map<String, Long> results = new HashMap();
+        Map<String, Map<String, Object>> blobstoreMap = (Map<String, Map<String, Object>>) workerState.getTopologyConf().get(Config.TOPOLOGY_BLOBSTORE_MAP);
+        if (blobstoreMap != null) {
+            String stormRoot = ConfigUtils.supervisorStormDistRoot(workerState.getTopologyConf(), workerState.getTopologyId());
+            for (Map.Entry<String, Map<String, Object>> entry : blobstoreMap.entrySet()) {
+                String keyName = entry.getKey();
+                String localFileName = keyName;
+                Map<String, Object> blobInfo = blobstoreMap.get(keyName);
+                if (blobInfo != null && blobInfo.containsKey("localname")) {
+                    localFileName = (String) blobInfo.get("localname");
+                }
+                String blobPath = new File(stormRoot, localFileName).getCanonicalPath();
+                File versionFile = new File(blobPath.replaceAll(".[0-9]+$", DEFAULT_BLOB_VERSION_SUFFIX));
+                results.put(localFileName, Utils.getVersionFromBlobVersionFile(versionFile));
+            }
+        }
+        return results;
+    }
+
+    public void updateBlobUpdates() throws IOException {
+        Map<String, Long> latestBlobVersions = getCurrentBlobVersions();
+        workerState.blobToLastKnownVersion.putAll(latestBlobVersions);
```

Comment:
We need some kind of synchronization here.  Perhaps we can just make `blobToLastKnownVersion` a ConcurrentMap.  That should be enough.